### DISCUSSION TITLE
Sept 20, 2022 Community Call

### DISCUSSION THREAD

**Discussion by expede** [ipvm-wg/spec#3] [Community Call]
Sign up: https://lu.ma/ipvm

# Suggested Agenda

_Let us know in the comments what you'd like to discuss! Questions about using UCAN, use cases in the wild, short demos, etc._

* @expede to more clearly delineate IPVM and Autocodec
* @softwareplumber describe ABI ideas related to XPath, XSLT, etc
* @RangerMauve exceptions, codecs, ADLs, byte loading
  * Quick @expede thoughts: exceptions are effects, go in job receipts
  * WASI article

# Video

# Chat Log

# Notes

**Comment by RangerMauve** [comment]
It'd be cool if we could talk about what sorts of ABIs we'd like to use.

Personally, I think it would be prudent to settle on ABIs between wasm and the host for the following (or to see if other folks are even into this approach):

- Memory management. How can we tell the host that the wasm module no longer needs a pointer
- What should exceptions look like (what exists)
- Codecs:
  - Providing bytes to a WASM module
  - APIs for the WASM module to construct an IPLD tree (new map, add this property, add a sub-map, etc)
  - Can this be done lazily, does it have to be all at once?
  - Do we use a similar pattern to [NodeBuilder](https://github.com/ipld/go-ipld-prime/blob/master/datamodel/nodeBuilder.go)
- ADLs
  - Use same APIs that WASM codecs use to represent a new Node
  - Use mirrored APIs for WASM to get a Node represented in the host
  - API for a WASM module to get a Node pointer by it's CID
- Spec for how WASM modules get loaded from IPFS
  - UnixFS for modules larger than a block?
  - Support ADLs like [FBL](https://ipld.io/specs/advanced-data-layouts/fbl/) as alternative to UnixFS?
  - What languages are folks willing to start implementing stuff in?

**Comment by RangerMauve** [comment]
Also would be cool to talk a bit about the WASI stuff Adin linked to

**Reply by RangerMauve** [reply]
https://devclass.com/2022/09/08/assemblyscript-project-wasi-damages-open-standards-and-the-web/

**Comment by nicobao** [comment]
Hi,
Thank you the call earlier, it was very interesting to listen.
I have a question, web assembly is promising and exciting, but docker - despite its limitations - is already a mature way to scale computing over a bunch of machines.
Apart from bacalhau, is IPVM working group also working towards this direction? 
On the FVM side - is it something that is being thought about as well? Or is there some technical limitations here that I am unaware of?
We could very well design an IPLD Data Model that would map the docker requirements and then run that externally, no?
